### DISCUSSION TITLE
Clarification about AlbumID3.artists

### DISCUSSION THREAD

**Discussion by deluan** [opensubsonic/open-subsonic-api#120] [General]
I just found what seems to be a conflict in OS spec: I'm adding `artists` to AlbumID3 in Navidrome, and [the spec](https://opensubsonic.netlify.app/docs/responses/albumid3/) says:

> Only the required ArtistID3 fields should be returned by default

And show this example:
```
    "artists": [
        {
            "id": "ar-1",
            "name": "Artist 1"
        },
        {
            "id": "ar-2",
            "name": "Artist 2"
        }
    ],
```

But at the same time, `ArtistID3` has some OS fields, `musicBrainzId`, `sortName`, `roles`. And [their specification](https://opensubsonic.netlify.app/docs/responses/artistid3/) says:

> All OpenSubsonic added fields are optionals. But if a server support a field it must return it with an empty / default value when not present in it’s database so that clients knows what the server supports.

This causes the `AlbumID3.artists`  to be marshaled as:
```
"artists": [
    {
        "id": "4Y4arYckrgDabKQyjrFYxh",
        "name": "TKKG",
        "musicBrainzId": "",
        "sortName": "",
        "roles": []
    }
]
```
with empty OS fields, conflicting with the AlbumID3 spec. 

Am I missing something? Maybe the text is not clear (at least for me)

**Comment by kgarner7** [comment]
I think that's just a problem with the example. The spec should be have mbid, sort name, and roles.

**Comment by epoupon** [comment]
Hmm for me this is just a more compact artist ref (id/name) to list the album artists. Otherwise we have to include/repeat a lot of info.
That is how it is currently implemented in lms (you can try the demo instance)

**Reply by epoupon** [reply]
And this is the same "light" Artist struct that is used in Contributor and in "artists"/"albumartists" in Child 

We added a note in the Contributor response to state that only the required fields should be returned

**Reply by deluan** [reply]
If that's the case, I think the object returned `AlbumID3.artists` is _not_ an `ArtistID3`, or else the documentation does not match. 

I agree we should not put all attributes of the artist there, so IMO a solution would be to define a _compact_ version of the `ArtistID3`, to be used in `AlbumID3.artists` and also in `Contributors`.

**Reply by jeffvli** [reply]
From a client perspective, returning the full `ArtistID3` would actually be better. Having only the id/name fields restricts its usability quite a bit.

For example, if I wanted to implement anything more than a basic text list of artists associated to an album, I would need to perform a fetch to `getArtist` for each artist to retrieve additional fields (or `getArtists`) which is not ideal.

**Reply by Tolriq** [reply]
Depends on the client and the call.

Typically returning full artists for a search3 on the albums would means that the same data would be send a millions times for no real gain.
For an app like Symfonium that would mean a sync duration that explodes and network data and battery for 0 gain.

Original spec already did not return all the data because it made sense. 

Here it's just a documentation issue for this special case, but the OS fields are not required so should not be returned. 
Just need a clarification of the doc and no need to create a fake other type, that would kinda break the original spec and complexify things for no real gain. (Clarification that for the calls that only returns required fields, the special OS rules that field that are supported should be returned is not applied, that clarify the issue and does not change anything to how things works currently)

Due to the amount of fields that will be added, the new endpoints will have a way to explicitly requests the fields that the client want for a lot of benefits solving those kind of issues.

**Reply by jeffvli** [reply]
~~You're right. I was thinking specifically of the `artists` property for `getAlbum`, but after looking at it again the documentation doesn't specify that only required fields are returned for that one so ignore my previous comment.~~ Actually, it seems like I got confused again on this. I would be right in saying that `getAlbum` that returns `AlbumID3` would only return the limited subset of artist data right? Since `AlbumID3` is what specifically states that it only returns required fields. This brings me down to the point below of wanting to create a new response type to differentiate between when to return the full `ArtistID3` vs a limited one.

> Just need a clarification of the doc and no need to create a fake other type, that would kinda break the original spec and complexify things for no real gain. (Clarification that for the calls that only returns required fields, the special OS rules that field that are supported should be returned is not applied, that clarify the issue and does not change anything to how things works currently)

I would actually want a new response type created (maybe `RelatedArtist` or `RelatedArtistID3` idk). It would make the docs more readable  rather than needing to dig down into the endpoint and then property description to find out that the field only returns required fields from the response type.

I ran into this issue with my [api client](https://github.com/audioling/open-subsonic-api-client) where I re-used the `ArtistID3` response type everywhere it was listed. While technically it's not *wrong* since the fields are optional, it would be better to explicitly state in a new response type that those fields would **never** be available under certain `artists` responses.

**Reply by deluan** [reply]
I agree this should be a new type (`ArtistID3Ref` or `RelatedArtistID3`?), as it simplifies the server side development, at least in Go where you can't easily add rules for JSON marshaling ("only omit theses fields in this situation"). I don't think it would be a breaking change as this is a field (`artists`) introduced by OS, and it doesn't really require any changes in any existing implementation (LMS, Symfonium, etc...)

Also, it would be very hard to add this to OpenAPI (same type, different fields requirements). FYI, @jeffvli is bravely working on having an OpenAPI spec for OpenSubsonic, and we should try to make things easier to get this spec out the door.

**Reply by Tolriq** [reply]
> FYI, @jeffvli is bravely working on having an OpenAPI spec for OpenSubsonic, and we should try to make things easier to get this spec out the door.

Again something related to here not done here :) You confirmed OpenAPI was not possible and so I started working on the JsonSchema for the new endpoints ....
Why not post here for those kind of things ? ....

Anyway if the changes are only tied to OS new things then yes a new type is OK, I thought there was some legacy cases too.

**Reply by jeffvli** [reply]
To be fair, the OpenAPI spec wasn't something I was specifically working on.

It's just a byproduct (auto-generated) of the API client that I built for my new project. We can still use JSONSchema if you find that it's more compatible with the spec, otherwise we could set up some Actions to migrate mine over to here.

**Reply by Tolriq** [reply]
As said in the other discussion I don't care, the only constraints is that it must work :)
So if it covers original API it must not result in auto generated code based on API spec to not match the actual result from actual servers.
I'm not an expert of both so as long as it works it's OK but AFAIK it was not possible to cover original spec with Open API and that generated issues to build something global.

@deluan probably have more details about what was not possible.

**Reply by deluan** [reply]
The issue was to get the same spec to cover both XML and JSON. IIRC, it was a pain to get it done correctly (with the correct wrappers for each response) in a way it does not break the available generators.

Unfortunately dropping support for XML would leave some notable clients out of the loop, like play:Sub, DSub and most notoriously, Supersonic.

**Reply by Tolriq** [reply]
Well we can drop XML for the new endpoints and have the OpenAPI only cover JSON.

That would not impact legacy clients at all since they would not use the OpenAPI, and if they want to migrate to OS they can switch to JSON.
Specially if the new endpoint only support JSON POST for complex endpoints parameters.

WDYT?

**Reply by deluan** [reply]
Not sure the value of having a spec that does not cover everything. Covering only JSON would not be useful for new servers wanting to use it to generate code. 

> Well we can drop XML for the new endpoints and have the OpenAPI only cover JSON.

If that means Using OpenAPI only for new endpoints could be confusing: It would force clients (and servers) to have multiple ways to handle the same API, if. they want to leverage code generation based on OpenAPI.

@jeffvli's client will probably generate a JSON only spec, it may be useful for clients, but servers will not be able to fully leverage it, if they want to serve a broader set of clients

I don't think there is a simple solution for this, that's why I stop pursuing it.

**Reply by Tolriq** [reply]
IMO I do not think an existing server would rewrite all it's code to migrate to OpenAPI auto generated code, but maybe I'm wrong on that one.

Any new server written from scratch, would probably not bother supporting older clients, when Navidrome will embrace multiple artists, people will want that even more.
There's more and more new clients on all platforms, I'm not sure it's worth to block new cool things for a couple of legacy clients that will disappear with time.

On Android, Google remove apps that do not update the target SDK, same on iOS (but later) and change rules. 
This will leave very few users installing APK (not possible on iOS).

TL;DR; let's not block the future if we can simplify every one life for a wrong reason.

**Reply by deluan** [reply]
> would probably not bother supporting older clients,

Yes, for DSub, but:
Supersonic is not an old client. c/c @dweymouth 
play:Sub is actively maintained. c/c @MichaelBechHansen 

> I do not think an existing server would rewrite all it's code to migrate to OpenAPI auto generated code, but maybe I'm wrong on that one.

Rewriting Subsonic support was the reason I started working on the OpenAPI spec... :)

Anyway, this is my 2c. Would love to hear others opinions on this.

**Reply by Tolriq** [reply]
Supersonic support Jellyfin so probably can migrate to Json in a couple of hours.

No idea about the iOS one :(

> Rewriting Subsonic support was the reason I started working on the OpenAPI spec... :)

And it's also the reason you could not finish ;)

Anyway to resume we both agree that a spec that does not cover everything is not really useful, so the compromise to have a full OpenAPI spec would be to only be JSON it seems.

If that's not a valid compromise, then I guess we keep JsonSchema for the new api endpoints at least for the documentation purpose if not the full auto client generation.

**Comment by deluan** [comment]
While we are at this, another thing that may need clarification. [`Child`](https://opensubsonic.netlify.app/docs/responses/child/) can be used for songs **AND** albums (and even artists!). So this section needs some better wording: 

| Field |  Type | Req. | OpenS. | Details |
| --- | --- | --- | --- | --- |
| `artists` | Array of [`ArtistID3`](../artistid3) | No | **Yes**    | The list of all song artists of the song. (Note: Only the required [`ArtistID3`](../artistid3) fields should be returned by default)|
| `displayArtist` | `string` | No |  **Yes**   | The single value display artist. |
| `albumArtists` | Array of [`ArtistID3`](../artistid3) | No | **Yes**    | The list of all album artists of the song. (Note: Only the required [`ArtistID3`](../artistid3) fields should be returned by default)|
| `displayAlbumArtist` | `string` | No |  **Yes**   | The single value display album artist. |
| `contributors` | Array of [`Contributor`](../contributor) | No | **Yes**    | The list of all contributor artists of the song. |

And when the `Child` is an album, `artists` would be a collection of track artists found in the album or should it be album artists and we remove the `albumArtists` field?

`displayArtist` does not make sense for album in this context (as it is a track artist), so are we going to add it as an empty field in this case? Or it would contain the value of DisplayAlbumArtist, in which case we don't need to duplicate it in `displayAlbumArtist`, right?

**Reply by Tolriq** [reply]
Child is a pain client side for now due to polymorphism based on a type value.
IMO we should not change the definition of the fields and for the displayArtist return empty and only use displayAlbumArtist.

**Reply by deluan** [reply]
It is a pain server side as well. 

> IMO we should not change the definition of the fields and for the displayArtist return empty and only use displayAlbumArtist.

But then it becomes inconsistent with what is done in AlbumID3, where we only have `displayArtist` and `artist`

I also checked `getAlbumList` from [LMS's demo server](https://lms-demo.poupon.dev/rest/getAlbumList?u=demo&p=7e5da62f-e4a2-f946-a790-9872352f82ae&f=json&v=1.8.0&c=NavidromeUI&type=random), and it returns the AlbumArtist in the `displayArtist`. @epoupon confirms?

Should we settle on having `displayArtist` when the Child is an Album, and fix the docs?

**Reply by Tolriq** [reply]
This is not really comparable, AlbumID3 fields are tied to it so a field X is X from Album.

Like field genre from track is not the same as genre from albums and would be different from artists if supported.
They are not named trackGenre, albumGenre because it's not needed.

Here the child can be a track or an album. In both cases displayAlbumArtist is consistent it's the display album artist.

What does it bring to the table to have an item with fields named X where X depends on the type to have different meanings?

The names here are clear and different by types, but even when the type change, the meaning of the fields is still fully valid.

**Reply by deluan** [reply]
> What does it bring to the table to have an item with fields named X where X depends on the type to have different meanings?

Hey, don't shoot the messenger! I agree and I and I even do it more explicitly in Navidrome: In albums I name artist as AlbumArtist, to avoid confusion.

But LMS already does this. Is @epoupon willing to change? I'm ok with both approaches.

**Reply by Tolriq** [reply]
As often misunderstood, let me tell again that I never shoot, I just ask the questions without chocolate coating :)
Not English native efficient simple sentences :)

**Reply by epoupon** [reply]
> > What does it bring to the table to have an item with fields named X where X depends on the type to have different meanings?
> 
> Hey, don't shoot the messenger! I agree and I and I even do it more explicitly in Navidrome: In albums I name artist as AlbumArtist, to avoid confusion.
> 
> But LMS already does this. Is @epoupon willing to change? I'm ok with both approaches.

No problem to change!

**Reply by deluan** [reply]
> > But LMS already does this. Is @epoupon willing to change? I'm ok with both approaches.
> 
> No problem to change!

@kgarner7 can you revise this in your "errata" PR, please?

**Reply by deluan** [reply]
So to confirm: for album Children we should populate only: `displayAlbumArtist` and `albumArtists`, leaving `displayArtist` and `artists` empty, right?

**Reply by kgarner7** [reply]
> > > But LMS already does this. Is @epoupon willing to change? I'm ok with both approaches.
> > 
> > 
> > No problem to change!
> 
> @kgarner7 can you revise this in your "errata" PR, please?

If `displayAlbumArtist` / `albumArtists` are only _sometimes_ populated (depending on the use), then it seems more correct to split up `Child` (at least for OS) into proper types (e.g. `AlbumChild`, `PodcastChild`, `SongChild`), rather than say "these fields are only required for one variant of the use".

I'm fine noting that these fields are not in fact required (and depend on use), but that seems more like a kludge.

**Reply by kgarner7** [reply]
Actually, on a related note, since `PodcastEpisode` extends child, which of the OS attributes are actually relevant? `artists`? `albumArtists`? I could see an argument for the other OS fields (with `displayComposer` _maybe_ having a different meaning)

**Reply by Tolriq** [reply]
> If displayAlbumArtist / albumArtists are only sometimes populated (depending on the use), then it seems more correct to split up Child (at least for OS) into proper types (e.g. AlbumChild, PodcastChild, SongChild), rather than say "these fields are only required for one variant of the use".

This is not possible for now as it would be polymorphism again, you can't really know what will come back before it comes back.

But things like Child will disappear in the new endpoints.

The fields are already not required, and it's actually the other ones that are sometimes not filled. Doc should just mention that displayArtist is for songs as is artists

**Reply by epoupon** [reply]
We also have mediaType to know if it is meant to be a song or an album. And we already have fields like musicBrainzId that have different meaning depending on the type.

**Reply by kgarner7** [reply]
1. `mediaType` doesn't necessarily help with podcasts (you would _also_ presumably want `type`)
2. Is `song/album/artist` the _only_ possible value for `mediaType` (no podcast/similar)? If so, then this should probably be represented as an enum, rather than a string.
3. If `podcast` is not a `mediaType`, then I would argue that `type` should also be required for OS servers.
4. None of the above really answers which of artist/album artist fields are relevant for podcasts.
